File: java/src/RagAgent.java
/**
 * RAG Agent — Java
 * Wires RagEngine (hybrid search, reranking, citations) + document loaders
 * into the multi-provider ReAct loop.
 *
 * export MODEL_PROVIDER=anthropic|deepseek|openai
 * export ANTHROPIC_API_KEY / DEEPSEEK_API_KEY / OPENAI_API_KEY
 * export VOYAGE_API_KEY=... QDRANT_URL=http://localhost:6333
 *
 * java -cp target/rag-ai-java-1.0.0.jar RagAgent --task "your question"
 * java -cp target/rag-ai-java-1.0.0.jar RagAgent --dir ./docs --task "your question"
 */

import java.nio.file.Path;
import java.util.*;

public class RagAgent {

    static final String RAG_SYSTEM_PROMPT =
        "You have access to a 'rag_search' tool that searches a knowledge base of ingested documents " +
        "using hybrid retrieval with reranking. Results include citation markers. " +
        "Use rag_search to find relevant context before answering factual questions. " +
        "Always search first, reason second, then answer. Cite retrieved passages.";

    static List<RagEngine.Document> defaultDocs() {
        return Arrays.asList(
            new RagEngine.Document("Claude is an AI assistant made by Anthropic, designed to be helpful, harmless, and honest.", Map.of("source", "anthropic_docs")),
            new RagEngine.Document("RAG (Retrieval-Augmented Generation) combines retrieval with generation to ground LLM answers in real documents.", Map.of("source", "rag_survey")),
            new RagEngine.Document("Qdrant is a high-performance vector database written in Rust, supporting cosine, dot product, and Euclidean distance.", Map.of("source", "qdrant_docs"))
        );
    }

    public static ReactLoop.ReactResult ingestAndRun(
        List<RagEngine.Document> documents,
        String task,
        String systemExtra,
        int maxIterations,
        boolean verbose
    ) throws Exception {
        // Init RAG engine
        RagEngine engine = new RagEngine("default");

        if (documents != null && !documents.isEmpty()) {
            int count = engine.ingest(documents);
            System.err.printf("[rag-agent] Ingested %d docs → %d chunks%n", documents.size(), count);
        }

        // Register rag_search tool in ReactLoop
        if (!ReactLoop.TOOLS.containsKey("rag_search")) {
            ReactLoop.registerTool(
                "rag_search",
                "Search the knowledge base for context relevant to a query.",
                ReactLoop.JSON.createObjectNode()
                    .<com.fasterxml.jackson.databind.node.ObjectNode>put("type", "object")
                    .set("properties", ReactLoop.JSON.createObjectNode()
                        .set("query", ReactLoop.JSON.createObjectNode().put("type", "string"))
                        .set("top_k", ReactLoop.JSON.createObjectNode().put("type", "integer"))),
                inputNode -> {
                    String query = inputNode.path("query").asText("");
                    int topK     = inputNode.path("top_k").asInt(5);
                    try {
                        return engine.retrieveAsContext(query, topK);
                    } catch (Exception e) {
                        return "rag_search error: " + e.getMessage();
                    }
                }
            );
        }

        String combined = systemExtra == null || systemExtra.isBlank()
            ? RAG_SYSTEM_PROMPT
            : RAG_SYSTEM_PROMPT + "\n\n" + systemExtra;

        return ReactLoop.runReact(task, combined, maxIterations, verbose ? step -> {
            switch (step.type()) {
                case "thought"     -> System.out.printf("%n💭 THOUGHT%n%s%n", step.content() == null ? "" : step.content().substring(0, Math.min(300, step.content().length())));
                case "action"      -> System.out.printf("%n⚡ ACTION  %s%n%s%n", step.tool(), step.input());
                case "observation" -> System.out.printf("%n%s OBSERVE%n%s%n", step.error() ? "❌" : "👁", step.content() == null ? "" : step.content().substring(0, Math.min(300, step.content().length())));
            }
        } : null);
    }

    public static void main(String[] args) throws Exception {
        String task    = null;
        String dir     = null;
        int maxIter    = 30;
        boolean verbose = true;

        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--task", "-task" -> { if (i + 1 < args.length) task = args[++i]; }
                case "--dir",  "-dir"  -> { if (i + 1 < args.length) dir  = args[++i]; }
                case "--max-iterations" -> { if (i + 1 < args.length) maxIter = Integer.parseInt(args[++i]); }
                case "--quiet"          -> verbose = false;
            }
        }

        if (task == null) task = "Explain how the ReAct pattern works and how it relates to RAG.";

        List<RagEngine.Document> documents;
        if (dir != null) {
            documents = Loaders.loadDirectory(Path.of(dir), true);
            if (documents.isEmpty()) {
                System.err.println("[rag-agent] No supported documents found in " + dir);
            }
        } else {
            documents = defaultDocs();
        }

        ReactLoop.ReactResult result = ingestAndRun(documents, task, null, maxIter, verbose);

        System.out.println("\n" + "=".repeat(64));
        System.out.println("FINAL ANSWER");
        System.out.println("=".repeat(64));
        System.out.println(result.answer());
        System.out.println("=".repeat(64));
        System.out.printf("Iterations: %d | Tokens: %d | Time: %dms%n",
            result.trace().iterations(), result.trace().totalTokens(), result.trace().elapsedMs());
    }
}

File: rust/src/rag_agent.rs
// RAG Agent — Rust
// Wires RAGEngine (hybrid search, reranking, citations) + document loaders
// into the multi-provider ReAct loop.
//
// export MODEL_PROVIDER=anthropic|deepseek|openai
// export ANTHROPIC_API_KEY / DEEPSEEK_API_KEY / OPENAI_API_KEY
// export VOYAGE_API_KEY=... QDRANT_URL=http://localhost:6333
//
// cargo run --bin run_agent -- agent --task "your question"
// cargo run --bin run_agent -- agent --dir ./docs --task "your question"

use std::env;
use std::path::Path;
use std::sync::Arc;

use anyhow::Result;
use serde_json::json;

use crate::{Document, RAGEngine, ToolRegistry, ReactOptions, run_react, load_directory, load_document};

const RAG_SYSTEM_PROMPT: &str = "You have access to a 'rag_search' tool that searches a knowledge base \
of ingested documents using hybrid retrieval with reranking. Results include citation markers. \
Use rag_search to find relevant context before answering factual questions. \
Always search first, reason second, then answer. Cite retrieved passages.";

fn default_docs() -> Vec<Document> {
    vec![
        Document {
            content: "Claude is an AI assistant made by Anthropic, designed to be helpful, harmless, and honest.".to_string(),
            metadata: [("source".to_string(), "anthropic_docs".to_string())].into(),
            doc_id: None,
        },
        Document {
            content: "RAG (Retrieval-Augmented Generation) combines retrieval with generation to ground LLM answers in real documents.".to_string(),
            metadata: [("source".to_string(), "rag_survey".to_string())].into(),
            doc_id: None,
        },
        Document {
            content: "Qdrant is a high-performance vector database written in Rust, supporting cosine, dot product, and Euclidean distance.".to_string(),
            metadata: [("source".to_string(), "qdrant_docs".to_string())].into(),
            doc_id: None,
        },
    ]
}

pub async fn run_rag_agent(
    docs: Vec<Document>,
    task: &str,
    system_extra: &str,
    max_iterations: usize,
    verbose: bool,
) -> Result<(String, crate::Trace)> {
    // Init RAG engine
    let engine = Arc::new(RAGEngine::new("default").await?);

    if !docs.is_empty() {
        let count = engine.ingest(docs).await?;
        eprintln!("[rag-agent] Ingested {} chunks", count);
    }

    // Build tool registry with rag_search wired in
    let mut registry = ToolRegistry::new();
    let engine_clone = Arc::clone(&engine);
    registry.register(
        "rag_search",
        "Search the knowledge base for context relevant to a query.",
        json!({
            "type": "object",
            "properties": {
                "query": { "type": "string" },
                "top_k": { "type": "integer" }
            },
            "required": ["query"]
        }),
        std::sync::Arc::new(move |input: serde_json::Value| {
            let engine = Arc::clone(&engine_clone);
            // Blocking call inside sync closure — safe for CLI usage
            let query  = input["query"].as_str().unwrap_or("").to_string();
            let top_k  = input["top_k"].as_u64().unwrap_or(5) as u64;
            let rt = tokio::runtime::Handle::current();
            std::thread::spawn(move || {
                rt.block_on(async move {
                    engine.retrieve_as_context(&query, top_k).await
                })
            }).join().map_err(|e| anyhow::anyhow!("thread panic: {:?}", e))?
        }),
    );

    let combined = if system_extra.is_empty() {
        RAG_SYSTEM_PROMPT.to_string()
    } else {
        format!("{}\n\n{}", RAG_SYSTEM_PROMPT, system_extra)
    };

    let on_step: Option<Box<dyn Fn(&crate::Step) + Send>> = if verbose {
        Some(Box::new(|step| match step {
            crate::Step::Thought { content }        => println!("\n💭 THOUGHT: {}", &content[..content.len().min(300)]),
            crate::Step::Action { tool, input }     => println!("\n⚡ ACTION: {}({})", tool, input.to_string().chars().take(200).collect::<String>()),
            crate::Step::Observation { content, error } => println!("\n{} OBSERVE: {}", if *error { "❌" } else { "👁" }, &content[..content.len().min(300)]),
            _ => {}
        }))
    } else {
        None
    };

    run_react(task, &registry, ReactOptions {
        system_extra:   combined,
        max_iterations,
        on_step,
    }).await
}

pub async fn run_agent_cli() -> Result<()> {
    let args: Vec<String> = env::args().skip(1).collect(); // skip binary name only
    let mut task = String::new();
    let mut dir  = String::new();
    let mut max_iter = 30usize;
    let mut verbose = true;

    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "--task" | "-task" => { if i + 1 < args.len() { task = args[i+1].clone(); i += 1; } }
            "--dir"  | "-dir"  => { if i + 1 < args.len() { dir  = args[i+1].clone(); i += 1; } }
            "--max-iterations"  => { if i + 1 < args.len() { max_iter = args[i+1].parse().unwrap_or(30); i += 1; } }
            "--quiet"           => { verbose = false; }
            _                   => {}
        }
        i += 1;
    }

    if task.is_empty() {
        task = "Explain how the ReAct pattern works and how it relates to RAG.".to_string();
    }

    let docs = if dir.is_empty() {
        default_docs()
    } else {
        load_directory(Path::new(&dir), true)
    };

    let (answer, trace) = run_rag_agent(docs, &task, "", max_iter, verbose).await?;

    println!("\n{}", "=".repeat(64));
    println!("FINAL ANSWER\n{}", "=".repeat(64));
    println!("{}", answer);
    println!("{}", "=".repeat(64));
    println!("Iterations: {} | Tokens: {} | Time: {}ms", trace.iterations, trace.total_tokens, trace.elapsed_ms);
    Ok(())
}
